### PR TITLE
Handle multiple popups

### LINKED ISSUE #77: Multiple "popups" on the same post
As far as I can tell, the insertion logic allows for either an inline or a modal popup, but not both on one post and not multiple inline popups. 

I think most sites have a setup like the following:
- Newsletter inline CTA
- One or more inline ad units (e.g. 3 and 6 paragraphs in)
- Newsletter/Donate modal

The solution IMO is the inline popups should not count towards the popup for the page. Let people insert as many inline "popups" as they want.

Further question: should people inject inline ads with this tool, or continue with SCAP?

Thoughts @jeffersonrabb, @thomasguillot?

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

A couple of changes to the logic and architecture, which result in the possibility of having multiple inline popups, and at most one overlaid popup:

- In editor, the "Sitewide Default" switch will be disabled if "inline" is chosen as placement.
- Some checks on the backend will prevent an inline popup from being marked as "sitewide default".
- Inline popups are inserted as shortcodes. This prevents an inline popup ending up inside another inline popup.
- All inline popups will be inserted into a post, unless there's category filtering. Any popup with a set category will only appear on a post with matching category.
- A non-inline popup will be displayed only if it's sitewide default, as it were (unless there's category filtering)
-  In admin's popup list view, a post state of "Inline" is added next to inline popups.

Closes #77 

### How to test the changes in this Pull Request:

1. Create a lot of popups - both inline and overlaid. Go wild.
2. Browse the site - all inline popups should be visible on posts.
3. Mark one of overlay (non-inline) popups as "Sitewide default".
4. Browse the site - apart from the inline popups, the overlaid sitewide popup should be displayed.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/class-newspack-popups-inserter.php [hunk 1]
```diff
@@ -13,85 +13,133 @@
...
 		$sitewide_default = get_option( Newspack_Popups::NEWSPACK_POPUPS_SITEWIDE_DEFAULT, null );
 		if ( $sitewide_default ) {
-			self::$popup = Newspack_Popups_Model::retrieve_popup_by_id( $sitewide_default );
-			if ( self::$popup ) {
-				return self::$popup;
+			$found_popup = Newspack_Popups_Model::retrieve_popup_by_id( $sitewide_default );
+			if (
+				$found_popup &&
+				// Prevent inline sitewide default from being added - all inline popups are there.
+				'inline' !== $found_popup['options']['placement']
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/class-newspack-popups-inserter.php]: Yes, and a safeguard against possible legacy inline && sitewide default popups getting duplicated

File: includes/class-newspack-popups-inserter.php [hunk 2]
```diff
@@ -13,85 +13,133 @@
...
-		$categories = get_the_category();
-		if ( $categories && count( $categories ) ) {
-			self::$popup = Newspack_Popups_Model::retrieve_popup( get_the_category() );
-			if ( self::$popup ) {
-				return self::$popup;
-			}
+		// Get all inline popups.
+		$inline_popups = Newspack_Popups_Model::retrieve_inline_popups();
+		if ( ! empty( $inline_popups ) ) {
+			$found_popups = array_merge(
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/class-newspack-popups-inserter.php]: That's true, I've refactored it so it's all more readable: 05b72ed

**Comment by adekbadek** [comment]
1. Cannot reproduce, the popup does not reappear for me :/
2. Should be fixed in 3d9eef44a24d5401b5c5c734d253c6678a00a30e

**Comment by jeffersonrabb** [comment]
> Cannot reproduce, the popup does not reappear for me :/

Here's the flow I'm following:

1) Delete all pop-ups.
2) Create Pop-up, Placement: Center, Trigger: Timer, Frequency: Once A Day. Check Sitewide Default. Publish and view the site Incognito. The pop-up should correctly appear on the homepage, then never again as you browse.
3) Create Pop-up, Placement: Center, Trigger: Timer, Frequency: Once A Day. Choose category "Events" Publish and view the site incognito. The first pop-up appears on the home, the events pop-up appears on 1st events page. Everything is correct.
4) Create a Pop-up, Placement: Inline, Frequency: Every Page. Publish, browse the site Incognito.
5) At this point, the Sitewide Default one should appear on every page.

**Comment by adekbadek** [comment]
Got it. This was because I haven't noticed that there can be only one `<amp-access>` component on a page. Fixed in fcc71a1899de88d07fe7caf43c549fe1d6213fd1 - now the `amp-access` has an array of access providers.

**Comment by adekbadek** [comment]
@jeffersonrabb I've just discovered that because of the use of shortcodes, inline popups are not previewable. Just a heads up that there will be changes to this PR.

File: includes/class-newspack-popups-inserter.php [hunk 3]
```diff
@@ -184,89 +199,77 @@ public static function insert_popup( $content = '', $popup = [] ) {
...
-		<?php
-		static::register_amp_scripts();
 	}
 
 	/**
 	 * Register and enqueue all required AMP scripts, if needed.
 	 */
 	public static function register_amp_scripts() {
-		if ( ! wp_script_is( 'amp-runtime', 'registered' ) ) {
+		if ( ! is_admin() && ! wp_script_is( 'amp-runtime', 'registered' ) ) {
```
Comment (jeffersonrabb) [inline includes/class-newspack-popups-inserter.php]: Why the `is_admin()` check here?

File: includes/class-newspack-popups-inserter.php [hunk 4]
```diff
@@ -293,26 +296,63 @@ public static function register_amp_scripts() {
...
 		return true;
 	}
+
+	/**
+	 * If Pop-up has categories, it should only be shown on posts/pages with those.
+	 *
+	 * @param object $popup The popup to assess.
+	 * @return bool Should popup be shown based on categories it has.
+	 */
+	public static function assess_categories_filter( $popup ) {
```
Comment (jeffersonrabb) [inline includes/class-newspack-popups-inserter.php]: Could you explain the need for this check? I haven't run it down thoroughly, but doesn't the logic in `retrieve_category_overlay_popup` already guarantee that there is category intersection between post and pop-up?

File: includes/class-newspack-popups-inserter.php [hunk 5]
```diff
@@ -184,89 +199,77 @@ public static function insert_popup( $content = '', $popup = [] ) {
...
-			if ( $position >= $precise_position ) {
-				$insertion_position = $position;
-				break;
-			}
+	public static function popup_shortcode( $atts = array() ) {
+		$previewed_popup_id = Newspack_Popups::previewed_popup_id();
+		if ($previewed_popup_id) {
+			return Newspack_Popups_Model::retrieve_preview_popup( $previewed_popup_id )['markup'];
+		} else {
+			return Newspack_Popups_Model::retrieve_popup_by_id( $atts['id'] )['markup'];
```
Comment (jeffersonrabb) [inline includes/class-newspack-popups-inserter.php]: Should this have a check that `$atts['id']` exists, and that the pop-up object returned actually has a `markup` param?

File: includes/class-newspack-popups-inserter.php (hunk 3 above)
Comment (adekbadek) [reply includes/class-newspack-popups-inserter.php:272]: Before, `register_amp_scripts` was called from `popup_access`, which [had this check](https://github.com/Automattic/newspack-popups/blob/98ddaa1c8d984045302134dfc099e1baa3ae0723/includes/class-newspack-popups-inserter.php#L223-L226). Now `register_amp_scripts` is called separately, so it has to check for itself. 
I've separated the two as part of refactoring.

File: includes/class-newspack-popups-inserter.php (hunk 5 above)
Comment (adekbadek) [reply includes/class-newspack-popups-inserter.php]: Fixed: 829adc938292df92b9347d2b137d3530cb5635d2

File: includes/class-newspack-popups-inserter.php [hunk 6]
```diff
@@ -13,129 +13,148 @@
...
-		// If nothing found, try for sitewide pop-up.
-		$sitewide_default = get_option( Newspack_Popups::NEWSPACK_POPUPS_SITEWIDE_DEFAULT, null );
-		if ( $sitewide_default ) {
-			self::$popup = Newspack_Popups_Model::retrieve_popup_by_id( $sitewide_default );
-			if ( self::$popup ) {
-				return self::$popup;
-			}
+		$popups_to_display = array_filter(
+			$popups_to_maybe_display,
+			[ 'self', 'should_display' ]
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/class-newspack-popups-inserter.php]: Fixed: 829adc938292df92b9347d2b137d3530cb5635d2

File: includes/class-newspack-popups-inserter.php [hunk 7]
```diff
@@ -184,89 +199,77 @@ public static function insert_popup( $content = '', $popup = [] ) {
...
-		$precise_position   = $total_length * $percentage;
-		$insertion_position = $total_length;
-		foreach ( $positions as $position ) {
-			if ( $position >= $precise_position ) {
-				$insertion_position = $position;
-				break;
-			}
+	public static function popup_shortcode( $atts = array() ) {
+		$previewed_popup_id = Newspack_Popups::previewed_popup_id();
+		if ($previewed_popup_id) {
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/class-newspack-popups-inserter.php]: Fixed: 829adc938292df92b9347d2b137d3530cb5635d2

File: includes/class-newspack-popups-inserter.php (hunk 4 above)
Comment (adekbadek) [reply includes/class-newspack-popups-inserter.php:339]: `retrieve_category_overlay_popup` does indeed guarantee category intersection between post and pop-up, but just for the _overlay_ popups. This check ensures that an inline popup with a set category will not be shown on a post with a different category.

**Pull request merged**